src/App.js

```diff
@@ -1,56 +1,203 @@
+  }
+
+
+  savePromotion = () => {
+    if (this.state.formData) {
+        let id = this.getPromotionId();
+        this.showAlertWithFormData();
+        this.setState(prevState => {return {  promotionData: { ...prevState.promotionData, [id ]: this.getNewOrUpdatedPromotion(id)}}})
+        this.setState({view: "Search"})
+    }
+  }
+
+
+  getNewOrUpdatedPromotion = (id) => {
+
+    let newPromotion = {
+      id: null,
+      devices: [],
+      ventures: [],
+      url: "",
+      name: "",
+      lastUpdatedTime: ""
+    }
+
+    newPromotion.id=id;
+    newPromotion.name=this.state.formData.name;
+    newPromotion.url=this.state.formData.url;
+    newPromotion.devices=[...this.state.formData.devices.keys()]
+    newPromotion.ventures=[...this.state.formData.ventures.keys()]
+    newPromotion.lastUpdatedTime=Date.now()
+
+    return newPromotion;
+  }
+
+
+  getPromotionId = () => {
+
+    let uuid4 = require('uuid4');
+    let id = null;
+    if(this.state.selectedPromotionId){
+      id = this.state.selectedPromotionId
+    }
+
+    if(this.state.formData.id){
+        id = this.state.formData.id
+    } else {
+      id = uuid4();
+    }
+
+    return id;
+  }
+
+
+  showAlertWithFormData = () => {
+    let savedData = `New Promotion Details:
+                    Name:\t ${this.state.formData.name} 
+                    Url:\t ${this.state.formData.url}
+                    Devices:\t ${getTruthyList(this.state.formData.devices )}
+                    Ventures:\t ${getTruthyList(this.state.formData.ventures)}`;
+    alert(savedData);
+  }
+
+
+  resetFormData = () => {
+    this.setState({
+      selectedPromotionId: null,
+      formData: {
+        devices: new Map(),
+        ventures: new Map(),
+        url: "",
+        name: ""
+      },
+    });
+  }
+
+
+  editBtnClick = (id) => {
+    this.setState({
+      selectedPromotionId: id,
+      view: "PromotionBuilder",
+      formData: this.setFormDataToSelectedPromotion(id)
+    });
+  }
+
+
+  setFormDataToSelectedPromotion = (selectedPromotionId) => {
+    let selectedPromotion = this.state.searchDisplay.find(item => item.id === selectedPromotionId);
+    let selectedPromotionFormData = {};
+    selectedPromotionFormData.devices = arrayToMap(selectedPromotion.devices)
+    selectedPromotionFormData.ventures = arrayToMap(selectedPromotion.ventures)
+    selectedPromotionFormData.url = selectedPromotion.url
+    selectedPromotionFormData.id = selectedPromotion.id
+    selectedPromotionFormData.name = selectedPromotion.name
+    return selectedPromotionFormData;
+  }
+
+
+  onMenuClick = (event) => {
     const selectedLink = event.currentTarget.className;
     let view;
-    switch(selectedLink){
+    switch (selectedLink) {
 
       case "search__link":
-          view = "Search";          
-          break;
+        view = "Search";
+        this.setState({searchTerm: ""})
+        this.searchPromotions();
```

Comment:
we are calling `searchPromotions` in `componentDidUpdate`. This invocation may be redundant.